### ISSUE TITLE
[onert] Introduce adam optimizer

### ISSUE THREAD

**Issue by ragmani** [Samsung/ONE#11403]
## What
Let's introduce Adam optimizer for training.

## Why
Adam optimizer is known as an optimizer that helps training very well. However, unlike SGD optimizer, Adam optimizer was not completed in #11014 because it requires additional gradient tensors for exponential moving average. 

## Tasks
- [x] Move SGD optimizer into train backend
  - #11511
  - #11648
- [x] Replace SGD kernel with an optimized kernel
  - #11656
  - #11658
- [x] Allocate memory for optimizers
  - #11659
- [x] Introduce Adam optimizer
  - #11588
- [x] Apply Adam optimizer
  - #11697

Draft: #11420

**Comment by ragmani** [comment]
I guess it's better to create and manage optimizers in `train` backend for the following reasons:
1. For managing tensors:
   - Some optimizers, such as Adam, need tensors that only they have access to
   - In onert, only tensor builders in backends create all tensors and allocate memory for tensors. 

2. For using accelerated kernels.
   - All optimizers have their own algorithms to adjust learning rate to optimize training. The algorithms have to be computed whenever each trainable tensor is updated. Therefore we need kernels that accelerate the algorithms.
   - In onert, only backends use kernels implemented in other libraries(e.g. cker, acl) to accelerate performance.

@Samsung/one_onert  Please give your opinion on the above.

**Comment by ragmani** [comment]
## Tensorflow Adam Eigen Kernel
https://github.com/tensorflow/tensorflow/blob/5e7d6faebab8758a1523810fbacd4a14b9803421/tensorflow/core/kernels/training_ops.cc#L3593-L3598
https://github.com/tensorflow/tensorflow/blob/5e7d6faebab8758a1523810fbacd4a14b9803421/tensorflow/core/kernels/training_ops.cc#L774-L845
```c++
template <typename T>
struct ApplyAdam<CPUDevice, T> : ApplyAdamNonCuda<CPUDevice, T> {};
```
```c++
using CPUDevice = Eigen::ThreadPoolDevice;
```

**Comment by ragmani** [comment]
## Tensorflow SGD Eigen Kernel
https://github.com/tensorflow/tensorflow/blob/5e7d6faebab8758a1523810fbacd4a14b9803421/tensorflow/core/kernels/training_ops.cc#L961-L999
https://github.com/tensorflow/tensorflow/blob/5e7d6faebab8758a1523810fbacd4a14b9803421/tensorflow/core/kernels/training_ops.cc#L45-L53

**Comment by ragmani** [comment]
## What is the member, `use_exclusive_lock_`?
```c++
template <typename Device, typename T>
class ApplyGradientDescentOp : public OpKernel {
 public:
  explicit ApplyGradientDescentOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
    OP_REQUIRES_OK(ctx, ctx->GetAttr("use_locking", &use_exclusive_lock_));
  }

  void Compute(OpKernelContext* ctx) override {
    const bool sparse = false;
    auto locks = MaybeLockVariableInputMutexesInOrder<Device, T>(
        ctx, use_exclusive_lock_, sparse, {0});
    Tensor var;
    OP_REQUIRES_OK(ctx, GetInputTensorFromVariable<Device, T>(
                            ctx, 0, use_exclusive_lock_, sparse, &var));

    OP_REQUIRES(
        ctx, var.IsInitialized(),
        errors::FailedPrecondition(
            "Attempting to use uninitialized variables: ", requested_input(0)));
    const Tensor& alpha = ctx->input(1);
    OP_REQUIRES(ctx, TensorShapeUtils::IsScalar(alpha.shape()),
                errors::InvalidArgument("alpha is not a scalar: ",
                                        alpha.shape().DebugString()));
    const Tensor& delta = ctx->input(2);
    OP_REQUIRES(
        ctx, var.shape().IsSameSize(delta.shape()),
        errors::InvalidArgument("var and delta do not have the same shape",
                                var.shape().DebugString(), " ",
                                delta.shape().DebugString()));

    const Device& device = ctx->template eigen_device<Device>();
    functor::ApplyGradientDescent<Device, T>()(
        device, var.flat<T>(), alpha.scalar<T>(), delta.flat<T>());

    MaybeForwardRefInputToRefOutput(ctx, 0, 0);
  }

 private:
  bool use_exclusive_lock_;
};
```

**Comment by ragmani** [comment]
- The result with adam optimizer
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train mnist_keras.circle --load_expected:raw out/train.output.1000.
bin --load_input:raw out/train.input.1000.bin --data_length 1000 --learning_rate 0.01 --optimizer 1
Model Expected Filename out/train.output.1000.bin
Model Input Filename out/train.input.1000.bin
Model Filename mnist_keras.circle
Epoch 1/5 - 1.262ms/step - loss: [0] 6.0912
Epoch 2/5 - 1.289ms/step - loss: [0] 0.1023
Epoch 3/5 - 1.196ms/step - loss: [0] 0.0954
Epoch 4/5 - 1.229ms/step - loss: [0] 0.0921
Epoch 5/5 - 1.227ms/step - loss: [0] 0.0907
===================================
MODEL_LOAD   takes 0.536 ms
PREPARE      takes 2.958 ms
EXECUTE      takes 196.250 ms
- MEAN     :  196.250 ms
- MAX      :  196.250 ms
- MIN      :  196.250 ms
- GEOMEAN  :  196.250 ms
===================================
```

**Comment by ragmani** [comment]
Done.
